activemodel/lib/active_model/type/model.rb

```diff
@@ -0,0 +1,61 @@
+# frozen_string_literal: true
+
+module ActiveModel
+  module Type
+    class Model < Value # :nodoc:
+      def initialize(**args)
+        @class_name = args.delete(:class_name)
+        @serializer = args.delete(:serializer) || ActiveSupport::JSON
+        super
+      end
+
+      def changed_in_place?(raw_old_value, value)
+        old_value = deserialize(raw_old_value)
+        old_value.attributes != value.attributes
+      end
+
+      def valid_value?(value)
+        return valid_hash?(value) if value.is_a?(Hash)
+
+        value.is_a?(klass)
+      end
+
+      def type
+        :model
+      end
+
+      def serializable?(value)
+        value.is_a?(klass)
+      end
+
+      def serialize(value)
+        serializer.encode(value.attributes_for_database)
+      end
+
+      def deserialize(value)
+        attributes = serializer.decode(value)
+        klass.new(attributes)
+      end
+
+      private
+        attr_reader :serializer
+
+        def valid_hash?(value)
+          value.keys.map(&:to_s).difference(klass.attribute_names).none?
+        end
+
+        def klass
+          @_model_type_class ||= @class_name.constantize
+        end
```

Comment:
What do you think about supporting both `String` and callable arguments for the `:class_name` option?

When combined with the [:collection type](https://github.com/rails/rails/pull/51421) (or even individually, on its own), the type of model to construct might depend on context-specific data in attributes (either the presence of specific keys or values for those keys like `:type`, `:role`, etc.).

```ruby
Membership = Struct.new(:name, :role)
AdminMembership = Class.new(Membership) do
  def name
    "#{super} (Admin)"
  end
end

class Group
  include ActiveModel::Model
  include ActiveModel::Attributes

  attribute :memberships, :collection, element_type: :model,
    class_name: ->(attrs) { attrs["role"] == "admin" ? "AdminMembership" : "Membership" }
end

member = {name: "A member", role: "member"}
admin = {name: "An admin", role: "admin"}
group = Group.new(memberships: [member, admin])
group.memberships.first.name # => "A member"
group.memberships.second.name # => "An admin (Admin)"
```

As an alternative to changing `:class_name` support, there could be an additional option like `:init_with` that defaults to `:new`, but could be configured to be any class level method or a callable that accepts the `Class` instance and the attributes `Hash`.

Does that kind of flexibility feel appropriate here?